File: src/kangaroo_reporter_ffi.mjs
import { toList } from "./gleam.mjs";

let cases = [];
let output = [];

export function append(value) {
  cases.push(value);
}

export function take() {
  const result = toList(cases);
  cases = [];
  return result;
}

export function append_output(caseName, stdout, stderr) {
  output.push([caseName, stdout, stderr]);
}

export function take_output() {
  const result = toList(output);
  output = [];
  return result;
}

File: src/kangaroo_reporter_ffi.erl
-module(kangaroo_reporter_ffi).
-export([append/1, take/0, append_output/3, take_output/0]).

-define(KEY, kangaroo_reporter_cases).
-define(OUTPUT_KEY, kangaroo_reporter_output).

append(Case) ->
    Existing = case get(?KEY) of undefined -> []; Cases -> Cases end,
    put(?KEY, [Case | Existing]),
    nil.

take() ->
    Cases = case get(?KEY) of undefined -> []; Existing -> lists:reverse(Existing) end,
    erase(?KEY),
    Cases.

append_output(CaseName, Stdout, Stderr) ->
    Existing = case get(?OUTPUT_KEY) of undefined -> []; Output -> Output end,
    put(?OUTPUT_KEY, [{CaseName, Stdout, Stderr} | Existing]),
    nil.

take_output() ->
    Output = case get(?OUTPUT_KEY) of
                 undefined -> [];
                 Existing -> lists:reverse(Existing)
             end,
    erase(?OUTPUT_KEY),
    Output.
